### PR TITLE
Layla - Edges - MediaRanker

### PR DESCRIPTION
**Pull request by lbristol88**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In the votes model I created two custom models. One of the models was the ability to know if a vote was already cast by a single user. This model prevents a user from voting for the same work more than once. 
Describe how you approached testing that model method. What edge cases did you come up with? | In the test I took the first vote that a user made on a specific work, called the method if they already voted, and had the expectation that it would return true if the vote was already completed. 
What are `session` and `flash`? What is the difference between them? | Flash and session are both hash like and send messages from the controller to the view. Session is similar to flash, but flash just shares the message once, where session will not go away after the next request-response cycle. Session is how a user can remain logged into a browser through the use of cookies.
Describe a controller filter you wrote. | In the works controller, I used the before action (before_action :find_work, only: [:show, :edit, :update, :destroy]) so that I wouldn't have to find the work in each time calling a method. 
What was one thing that you gained more clarity on through this assignment? | The interaction between routes and controllers. Also using flash.
What is the Heroku URL of your deployed application? | https://layla-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? |

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      redirect_to users_path
```
Comment (droberts-sea) [inline app/controllers/users_controller.rb:14]: I don't think you need `new` and `create` in both the users and sessions controllers - having this functionality in one place or the other makes more sense.

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      @books = Work.works_by_category("book")
+      @albums = Work.works_by_category("album")
+      @movies = Work.works_by_category("movie")
+    end
+
+    def homepage
+      @books = Work.top_works_by_category("book")
+      @albums = Work.top_works_by_category("album")
+      @movies = Work.top_works_by_category("movie")
+      @top_work = Work.top_work
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:15]: Your model methods make this controller action very clean. Good work!

File: app/models/vote.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  belongs_to :user
+
+  def self.votes_by_work(work_id)
+    return votes = Vote.where(work_id: work_id)
+  end
+
+  def self.already_voted?(work_id, user_id)
+    vote = Vote.where(work_id: work_id, user_id: user_id)
+
+    return vote.any?
```
Comment (droberts-sea) [inline app/models/vote.rb:12]: I like that you implemented this as a model method - good instincts. If you wanted to go even further, you could have used a `uniqueness` validation with the `scope` parameter to do this automatically.

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  validates :creator, :category, :description, :publication_year, presence: true
+
+  def self.works_by_category(category)
+    return Work.where(category: category)
+  end
+
+  def self.top_works_by_category(category)
+    works = self.works_by_category(category)
+    works = works.sort_by {|work| work.votes.count}.reverse!
+   return works[0..9]
```
Comment (droberts-sea) [inline app/models/work.rb:14]: I like how these methods build on each other - good functional decomposition.

File: app/views/works/homepage.html.erb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    <%= "#{@top_work.votes.count} votes - #{@top_work.description}." %>
+  </p>
+</section>
+</section>
+</section>
+
+<section class="top-ten__container">
+ <% [[@books, "Books"], [@albums, "Albums"], [@movies, "Movies"]].each do |works, category| %>
+ <section class="top-ten__list-container">
+   <h2 class="top-ten__header">
```
Comment (droberts-sea) [inline app/views/works/homepage.html.erb:15]: Nice work keeping this DRY using a loop!

File: config/routes.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+
+
+  root 'works#homepage'
+
+  resources :votes
+  resources :users
+  resources :works
```
Comment (droberts-sea) [inline config/routes.rb:8]: I don't think you need all 7 RESTful routes for users and votes. You should only be generating the routes you need.

File: test/models/vote_test.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    it "has a user" do
+      v = Vote.first
+      user = v.user
+      expect(user).must_be_instance_of User
+    end
+
+    it "a user can only vote on the same work once" do
+
+      v = Vote.first
```
Comment (droberts-sea) [inline test/models/vote_test.rb:20]: I would like to see a few more tests for the `already_voted?` method. Specific test cases I'd want to see:

- A user can have votes for two different works
- A work can have votes from two different users
- A user cannot vote for the same work twice

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      @work.title = nil
+      is_valid = @work.valid?
+
+      expect( is_valid ).must_equal false
+      expect( @work.errors.messages ).must_include :title
+    end
+
+    it 'is invalid with a non-unique title' do
+      @work.title = Work.first.title
+      is_valid = @work.valid?
```
Comment (droberts-sea) [inline test/models/work_test.rb:46]: There's a lot of interesting test cases for your custom `Work` methods missing here. For `top_work`, I would wonder:

- What happens if there are no works?
- What happens if there are works but no votes?
- What happens if two works have the same number of votes?

Similarly for your category and top ten methods, I would ask:

- What if there are no works of that category?
- What if there are less than 10 works?
- What if there's a tie for last place, e.g. works 9, 10 and 11 all have 0 votes?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | some extra routes generated
Views are well-organized (DRY, use of semantic HTML, use of partials) | yes
Errors are reported to the user | some - no validation failures reported
Business logic lives in the models | yes - great work
Models are thoroughly tested, including relations, validations and any custom logic | many cases missing - see inline
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** | Good work overall. You've done a great job replicating the functionality of the demo site, and it's clear to me that most of the learning goals for this assignment were met. The one place where I see room for improvement is testing - make sure to review my inline comments, and practice this intentionally on future projects. I would be happy to sit down and review your test coverage before a project is due if that would help. Other than that I'm quite happy with this submission - keep up the hard work!
